rrdd/design/plugin-protocol-v3.md

```diff
@@ -0,0 +1,81 @@
+---
+title: RRDD plugin protocol v3
+layout: default
+design_doc: true
+revision: 1
+status: proposed
+revision_history:
+- revision_number: 1
+  description: Initial version
+---
+
+Motivation
+----------
+
+rrdd plugins protocol v2 report datasources via shared-memory file, however it
+has various limitations :
+ - metrics are unique by their names, thus it is not possible cannot have
+  several metrics that shares a same name (e.g vCPU usage per vm)
+ - only number metrics are supported, for example we can't expose string
+  metrics (e.g CPU Model)
+
+Therefore, it implies various limitations on plugins and limits
+[OpenMetrics](https://openmetrics.io/) support for the metrics daemon.
+
+Moreover, it may not be practical for plugin developpers and parser implementations :
+ - json implementations may not keep insersion order on maps, which can cause
+   issues to expose datasource values as it is sensitive to the order of the metadata map
+ - header length is not constant and depends on datasource count, which complicates parsing
+ - it still requires a quite advanced parser to convert between bytes and numbers according to metadata
+
+A simpler protocol is proposed, based on OpenMetrics binary format to ease plugin and parser implementations.
+
+Protocol V3
+-----------
+
+For this protocol, we still use a shared-memory file, but significantly change the structure of the file.
+
+|     value      |        bits        | format |                             notes
+| -------------- | ------------------ | ------ | ------------------------------------------------------------
+| header string  |      12*8=96       | string | "OPENMETRICS1" which is one byte longer than "DATASOURCES", intentionally made at 12 bytes for alignment purposes
+| data checksum  |         32         | uint32 | Checksum of the concatenation of the rest of the header (from timestamp) and the payload data
+| timestamp      |         64         | uint64 | Unix epoch
+| payload length |         32         | uint32 | Payload length
+| payload data   | 8*(payload length) | binary | OpenMetrics encoded metrics data (protocol-buffers format)
```

Comment:
The format is a container format. It could be helpful to add  a field that announces the format of the payload, in particular if we want to permit different versions in the payload. Otherwise we would have to encode this in the header string - which is still possible. If the OpemMetrics header is sufficiently identifying its version and format, it might not be needed.

Reply:
OpenMetrics protobuf format doesn't provide versioning.
It is possible to add a version field ?

The only problem is on how to encode the version, we can still use a version table that maps a version "integer" into a OpenMetrics version. The implementation will have to use this field to check compatibility. However, it is unclear how to choose what version to target (plugins should always use latest one available ?)

Reply:
"plugins should always use latest one available" is not a sustainable position and any client reading the data should be able to rely on version information to know if it is compatible. Otherwise a client would have to parse the data and on failure decide that it is incompatible. So I suggest to make a client's life easy by announcing formats or versions ahead of the data if the data by itself has no header that makes this easy.

Reply:
Like a new RPC function for plugin registering using protocol v3 (like Plugin.Local.RegisterV3 ?) that has a field "versions" with a list of OpenMetrics versions, and then the server answer with either "no version supported" or with the OpenMetrics version that must be used.